File: leetcode/editor/cn/349_两个数组的交集_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when349_thenSuccess(t *testing.T) {
	assert.Equal(t, 1, 1, "shouldEqual")
}

//leetcode submit region begin(Prohibit modification and deletion)
func intersection(nums1 []int, nums2 []int) []int {
	var ans []int
	if len(nums1) > len(nums2) {
		nums1, nums2 = nums2, nums1
	}
	set := map[int]int{}
	for _, num := range nums1 {
		set[num] = 1
	}
	// 思路：遍历 nums2 时更新 ans
	for _, num := range nums2 {
		// TODO set[notExistsKey] = 0
		if set[num] > 0 {
			ans = append(ans, num)
			set[num] = 0
		}
	}
	return ans
}

func intersection1(nums1 []int, nums2 []int) []int {
	if len(nums1) > len(nums2) {
		nums1, nums2 = nums2, nums1
	}
	set := map[int]int{}
	for _, num := range nums1 {
		set[num] = 1
	}
	for _, num := range nums2 {
		if cnt, ok := set[num]; ok {
			set[num] = cnt + 1
		}
	}
	var ans []int
	for k, v := range set {
		if v > 1 {
			ans = append(ans, k)
		}
	}
	return ans
}

//leetcode submit region end(Prohibit modification and deletion)

File: leetcode/editor/cn/349_两个数组的交集_test.py
import unittest
from typing import List


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        self.assertTrue(True, "ShouldBeTrue")


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:

    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        ans = []
        base_cnt = {}
        for x in nums1:
            base_cnt[x] = 1
        for x in nums2:
            if x in base_cnt:
                del base_cnt[x]
                ans.append(x)
        return ans

    def intersection1(self, nums1: List[int], nums2: List[int]) -> List[int]:
        base_cnt = {}
        for x in nums1:
            base_cnt[x] = 1
        for x in nums2:
            if x in base_cnt:
                base_cnt[x] += 1
        ans = []
        # TODO 遍历 map
        for k, v in base_cnt.items():
            if v > 1:
                ans.append(k)
        return ans


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()
